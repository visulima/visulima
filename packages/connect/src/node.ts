import type { IncomingMessage, ServerResponse } from "node:http";
import type { AnyZodObject, ZodObject } from "zod";

import withZod from "./adapter/with-zod";
import type { Route } from "./router";
import { Router } from "./router";
import type {
    FindResult,
    FunctionLike,
    HandlerOptions,
    HttpMethod,
    Nextable,
    RouteMatch,
    RouteShortcutMethod,
    RoutesExtendedRequestHandler,
    ValueOrPromise,
} from "./types";

const onNoMatch = async (request: IncomingMessage, response: ServerResponse) => {
    response.statusCode = 404;
    response.end(request.method === "HEAD" ? undefined : `Route ${request.method} ${request.url} not found`);
};

const onError = async (error: unknown, _request: IncomingMessage, response: ServerResponse) => {
    response.statusCode = 500;

    console.error(error);

    response.end("Internal Server Error");
};

export const getPathname = (url: string): string => {
    const queryIndex = url.indexOf("?");

    return queryIndex === -1 ? url : url.slice(0, Math.max(0, queryIndex));
};

export type RequestHandler<Request extends IncomingMessage, Response extends ServerResponse> = (request: Request, response: Response) => ValueOrPromise<void>;

export class NodeRouter<
    Request extends IncomingMessage = IncomingMessage,
    Response extends ServerResponse = ServerResponse,
    Schema extends AnyZodObject = ZodObject<never>,
> {
    private readonly onError: (
        error: unknown,
        ...arguments_: Parameters<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>>
    ) => ReturnType<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>>;

    private readonly onNoMatch: RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>;

    private router = new Router<RequestHandler<Request, Response>>();

    public all: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "");

    public delete: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "DELETE");

    public get: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "GET");

    public head: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "HEAD");

    public patch: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "PATCH");

    public post: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "POST");

    public put: RouteShortcutMethod<this, Schema, RequestHandler<Request, Response>> = this.add.bind(this, "PUT");

    public constructor(options: HandlerOptions<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>> = {}) {
        this.onNoMatch = options.onNoMatch ?? onNoMatch;
        this.onError = options.onError ?? onError;
    }

    private add(
        method: HttpMethod | "",
        routeOrFunction: Nextable<RequestHandler<Request, Response>> | RouteMatch,
        zodOrRouteOrFunction?: Nextable<RequestHandler<Request, Response>> | RouteMatch | Schema,
        ...fns: Nextable<RequestHandler<Request, Response>>[]
    ) {
        if (typeof routeOrFunction === "string" && typeof zodOrRouteOrFunction === "function") {
            // eslint-disable-next-line no-param-reassign
            fns = [zodOrRouteOrFunction];
        } else if (typeof zodOrRouteOrFunction === "object") {
            if (typeof routeOrFunction === "function") {
                // eslint-disable-next-line no-param-reassign
                fns = [withZod<Request, Response, Nextable<RequestHandler<Request, Response>>, Schema>(zodOrRouteOrFunction as Schema, routeOrFunction)];
            } else {
                // eslint-disable-next-line no-param-reassign
                fns = fns.map((function_) =>
                    withZod<Request, Response, Nextable<RequestHandler<Request, Response>>, Schema>(zodOrRouteOrFunction as Schema, function_),
                );
            }
        } else if (typeof zodOrRouteOrFunction === "function") {
            // eslint-disable-next-line no-param-reassign
            fns = [zodOrRouteOrFunction];
        }

        this.router.add(method, routeOrFunction, ...fns);

        return this;
    }

    // eslint-disable-next-line class-methods-use-this
    private prepareRequest(request: Request & { params?: Record<string, unknown> }, findResult: FindResult<RequestHandler<Request, Response>>) {
        request.params = {
            ...findResult.params,
            ...request.params, // original params will take precedence
        };
    }

    public clone(): NodeRouter<Request, Response, Schema> {
        const r = new NodeRouter<Request, Response, Schema>({ onError: this.onError, onNoMatch: this.onNoMatch });

        r.router = this.router.clone();

        return r;
    }

    public handler(): (request: Request, response: Response) => Promise<void> {
        const { routes } = this.router as Router<FunctionLike>;

        return async (request: Request, response: Response) => {
            // eslint-disable-next-line unicorn/no-array-callback-reference,unicorn/no-array-method-this-argument
            const result = this.router.find(request.method as HttpMethod, getPathname(request.url as string));

            this.prepareRequest(request, result);

            try {
                await (result.fns.length === 0 || result.middleOnly ? this.onNoMatch(request, response, routes) : Router.exec(result.fns, request, response));
            } catch (error) {
                await this.onError(error, request, response, routes);
            }
        };
    }

    public async run(request: Request, response: Response): Promise<unknown> {
        // eslint-disable-next-line unicorn/no-array-callback-reference,unicorn/no-array-method-this-argument
        const result = this.router.find(request.method as HttpMethod, getPathname(request.url as string));

        if (result.fns.length === 0) {
            return;
        }

        this.prepareRequest(request, result);

        // eslint-disable-next-line consistent-return
        return await Router.exec(result.fns, request, response);
    }

    public use(
        base: Nextable<RequestHandler<Request, Response>> | NodeRouter<Request, Response, Schema> | RouteMatch,
        ...fns: (Nextable<RequestHandler<Request, Response>> | NodeRouter<Request, Response, Schema>)[]
    ): this {
        if (typeof base === "function" || base instanceof NodeRouter) {
            fns.unshift(base);
            // eslint-disable-next-line no-param-reassign
            base = "/";
        }
        this.router.use(base, ...fns.map((function_) => (function_ instanceof NodeRouter ? function_.router : function_)));

        return this;
    }
}

export const createRouter = <
    Request extends IncomingMessage,
    Response extends ServerResponse,
    Schema extends AnyZodObject = ZodObject<{ body?: AnyZodObject; headers?: AnyZodObject; query?: AnyZodObject }>,
>(
    options: HandlerOptions<RoutesExtendedRequestHandler<Request, Response, Response, Route<Nextable<FunctionLike>>[]>> = {},
): NodeRouter<Request, Response, Schema> => new NodeRouter<Request, Response, Schema>(options);
