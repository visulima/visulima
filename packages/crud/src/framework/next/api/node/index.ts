import type { NextApiRequest, NextApiResponse } from "next";

import baseHandler from "../../../../base-crud-handler";
import type {
    Adapter, ExecuteHandler, HandlerOptions, ParsedQueryParameters,
} from "../../../../types";

async function handler<
    T,
    Q extends ParsedQueryParameters = any,
    R extends NextApiRequest = NextApiRequest,
    Response extends NextApiResponse = NextApiResponse,
    M extends string = string,
>(adapter: Adapter<T, Q>, options?: HandlerOptions<M, R, Response>): Promise<ExecuteHandler<R, Response>> {
    return baseHandler<R, Response, T, Q, M>(
        async (response, responseConfig) => {
            response.status(responseConfig.status).send(responseConfig.data);
        },
        async (response) => {
            response.end();
        },
        adapter,
        options,
    );
}

export default handler;
