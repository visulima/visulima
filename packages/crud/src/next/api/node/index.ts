import type { NextApiRequest, NextApiResponse } from "next";

import baseHandler from "../../../base-crud-handler";
import type { Adapter, ExecuteHandler, HandlerOptions, ParsedQueryParameters } from "../../../types.d";

const handler = async <
    T,
    Q extends ParsedQueryParameters = any,
    R extends NextApiRequest = NextApiRequest,
    Response extends NextApiResponse = NextApiResponse,
    M extends string = string,
>(
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, Response>> =>
    await baseHandler<R, Response, T, Q, M>(
        async (response, responseConfig) => {
            response.status(responseConfig.status).send(responseConfig.data);
        },
        async (response) => {
            response.end();
        },
        adapter,
        options,
    );

export default handler;
