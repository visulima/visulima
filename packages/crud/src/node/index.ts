import type { IncomingMessage, ServerResponse } from "node:http";

import baseHandler from "../base-crud-handler";
import type {
    Adapter, ExecuteHandler, HandlerOptions, ParsedQueryParameters,
} from "../types.d";

async function handler<R extends IncomingMessage, Response extends ServerResponse, T, Q extends ParsedQueryParameters = any, M extends string = string>(
    adapter: Adapter<T, Q>,
    options?: HandlerOptions<M>,
): Promise<ExecuteHandler<R, Response>> {
    return baseHandler<R, Response, T, Q, M>(
        async (response, responseConfig) => {
            response.setHeader("content-type", "application/json; charset=utf-8");

            response.statusCode = responseConfig.status;
            response.end(JSON.stringify(responseConfig.data, null, 2));
        },
        async (response) => {
            (response as Response).end();
        },
        adapter,
        options,
    );
}

export default handler;
