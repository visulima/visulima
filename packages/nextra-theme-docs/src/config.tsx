export { useConfig } from "./contexts";

export { type PartialDocumentsThemeConfig as DocumentationThemeConfig } from "./theme/theme-schema";
