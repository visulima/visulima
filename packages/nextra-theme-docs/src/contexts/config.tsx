import { ThemeProvider } from "next-themes";
import { metaSchema } from "nextra/normalize-pages";
import type { FrontMatter, PageMapItem, PageOpts } from "nextra/types";
import type { ReactElement, ReactNode } from "react";
import { createContext, useContext, useMemo, useState } from "react";
import type { ZodError } from "zod";
import { z } from "zod";

import { DEEP_OBJECT_KEYS, DEFAULT_THEME } from "../constants/default-theme-config";
import type { DocumentationThemeConfig } from "../theme/theme-schema";
import { themeSchema } from "../theme/theme-schema";
import type { Context } from "../types";
import { MenuProvider } from "./menu";

const ConfigContext = createContext<Config>({
    description: "",
    frontMatter: {},
    title: "",
    ...DEFAULT_THEME,
});

let theme: DocumentationThemeConfig;
let isValidated = false;

const extendedMetaSchema = metaSchema.or(
    // eslint-disable-next-line zod/require-strict
    z
        .object({
            description: z.string().optional(),
            icon: z.string().optional(),
            // eslint-disable-next-line zod/require-strict
            theme: z
                .object({
                    prose: z.boolean().optional(),
                })
                .optional(),
        })
        .optional(),
);

const normalizeZodMessage = (error: unknown): string =>
    (error as ZodError).issues
        .flatMap((issue) => {
            const themePath = issue.path.length > 0 && `Path: "${issue.path.join(".")}"`;
            const unionErrors = "unionErrors" in issue ? issue.unionErrors.map((data) => normalizeZodMessage(data)) : [];
            return [[issue.message, themePath].filter(Boolean).join(". "), ...unionErrors];
        })
        .join("\n");

const validateMeta = (pageMap: PageMapItem[]) => {
    pageMap.forEach((pageMapItem) => {
        if (pageMapItem.kind === "Meta") {
            Object.entries(pageMapItem.data).forEach(([key, value]) => {
                try {
                    extendedMetaSchema.parse(value);
                } catch (error) {
                    // eslint-disable-next-line no-console
                    console.error(`[nextra-theme-docs] Error validating _meta.json file for "${key}" property.\n\n${normalizeZodMessage(error)}`);
                }
            });
        } else if (pageMapItem.kind === "Folder") {
            validateMeta(pageMapItem.children);
        }
    });
};

export const useConfig = <FrontMatterType = FrontMatter,>(): Config<FrontMatterType> =>
    // @ts-expect-error TODO: fix Type 'Config<{ [key: string]: any; }>' is not assignable to type 'Config<FrontMatterType>'.
    useContext<Config<FrontMatterType>>(ConfigContext);

export const ConfigProvider = ({ children, value: { pageOpts, themeConfig } }: { children: ReactNode; value: Context }): ReactElement => {
    const [menu, setMenu] = useState(false);

    // Merge only on first load
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
    theme ||= {
        ...DEFAULT_THEME,
        ...Object.fromEntries(
            Object.entries(themeConfig).map(([key, value]) => [
                key,
                value && typeof value === "object" && DEEP_OBJECT_KEYS.includes(key)
                    ? // @ts-expect-error -- key has always object value
                      // eslint-disable-next-line security/detect-object-injection
                      { ...DEFAULT_THEME[key], ...(value as object) }
                    : value,
            ]),
        ),
    };

    if (process.env.NODE_ENV !== "production" && !isValidated) {
        try {
            themeSchema.parse(theme);
        } catch (error) {
            // eslint-disable-next-line no-console
            console.error(`[nextra-theme-docs] Error validating theme config file.\n\n${normalizeZodMessage(error)}`);
        }

        validateMeta(pageOpts.pageMap);

        isValidated = true;
    }

    const extendedConfig: Config = useMemo(() => {
        return {
            ...theme,
            description: pageOpts.description,
            flexsearch: pageOpts.flexsearch,
            frontMatter: pageOpts.frontMatter,
            title: pageOpts.title,
        };
    }, [pageOpts]);

    const { darkMode, nextThemes } = extendedConfig;
    const forcedTheme = darkMode ? nextThemes.forcedTheme : "light";

    const menuValue = useMemo(() => {
        return { menu, setMenu };
    }, [menu, setMenu]);

    return (
        <ThemeProvider
            attribute={nextThemes.attribute ?? "class"}
            defaultTheme={nextThemes.defaultTheme}
            disableTransitionOnChange
            forcedTheme={forcedTheme}
            storageKey={nextThemes.storageKey}
        >
            <ConfigContext.Provider value={extendedConfig}>
                <MenuProvider value={menuValue}>{children}</MenuProvider>
            </ConfigContext.Provider>
        </ThemeProvider>
    );
};

export type Config<FrontMatterType = FrontMatter> = DocumentationThemeConfig &
    Pick<PageOpts<FrontMatterType>, "flexsearch" | "frontMatter" | "title"> & { description?: string };
