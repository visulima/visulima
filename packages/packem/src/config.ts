import type { BuildConfig, BuildPreset } from "./types";

export const defineConfig = (config: BuildConfig | BuildConfig[]): BuildConfig[] => (Array.isArray(config) ? config : [config]).filter(Boolean);

export const definePreset = (preset: BuildPreset): BuildPreset => preset;
