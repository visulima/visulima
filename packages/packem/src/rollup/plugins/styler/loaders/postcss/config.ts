import path from "node:path";

import { cosmiconfig } from "cosmiconfig";
import type { AcceptedPlugin, Parser, PluginCreator, Stringifier, Syntax } from "postcss";

import type { PostCSSConfigLoaderOptions } from "../../types";
import { ensurePCSSOption, ensurePCSSPlugins } from "../../utils/options";

interface Options {
    parser?: Parser;
    stringifier?: Stringifier;
    syntax?: Syntax;
}

interface Config extends Options {
    plugins?: (PluginCreator<unknown> | string)[] | Record<string, Record<string, unknown>>;
}

interface Result {
    options: Options;
    plugins: AcceptedPlugin[];
}

export default async function (id: string, config: PostCSSConfigLoaderOptions | false): Promise<Result> {
    if (!config) {
        return { options: {}, plugins: [] };
    }

    const { base, dir, ext } = path.parse(id);

    type Found = { config: Config | ((context: Record<string, unknown>) => Config); isEmpty?: boolean };
    const searchPath = config.path ? path.resolve(config.path) : dir;
    const found: Found | null = await cosmiconfig("postcss").search(searchPath);

    if (!found || found.isEmpty) {
        return { options: {}, plugins: [] };
    }

    const { parser, plugins, stringifier, syntax } =
        typeof found.config === "function"
            ? found.config({
                  cwd: process.cwd(),
                  env: process.env.NODE_ENV ?? "development",
                  file: { basename: base, dirname: dir, extname: ext },
                  options: config.ctx ?? {},
              })
            : found.config;

    const result: Result = { options: {}, plugins: ensurePCSSPlugins(plugins) };

    if (parser) {
        result.options.parser = ensurePCSSOption(parser, "parser");
    }

    if (syntax) {
        result.options.syntax = ensurePCSSOption(syntax, "syntax");
    }

    if (stringifier) {
        result.options.stringifier = ensurePCSSOption(stringifier, "stringifier");
    }

    return result;
}
