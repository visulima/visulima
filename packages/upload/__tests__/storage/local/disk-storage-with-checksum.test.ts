import { join } from "node:path";
import { Readable } from "node:stream";

import { describe, expect, it, jest } from "@jest/globals";
import { createRequest } from "node-mocks-http";
import { describe, expect, it } from "vitest";

import DiskStorageWithChecksum from "../../../src/storage/local/disk-storage-with-checksum";
import { metafile, storageOptions, testRoot } from "../../__helpers__/config";

jest.mock("fs/promises", () => {
    const process = require("node:process");

    process.chdir("/");

    const { fs } = require("memfs");

    return fs.promises;
});

jest.mock("fs", () => {
    const process = require("node:process");

    process.chdir("/");

    const { fs } = require("memfs");

    return fs;
});

// eslint-disable-next-line radar/no-identical-functions
jest.mock("node:fs", () => {
    const process = require("node:process");

    process.chdir("/");

    const { fs } = require("memfs");

    return {
        __esModule: true,
        ...fs,
    };
});

const directory = join(testRoot, "disk-storage", "checksum");

describe(DiskStorageWithChecksum, () => {
    jest.useFakeTimers({ doNotFake: ["setTimeout"] }).setSystemTime(new Date("2022-02-02"));

    const options = { ...storageOptions, checksum: "sha1" as const, directory };
    const request = createRequest();

    it("should support checksum resume from fs", async () => {
        const storage = new DiskStorageWithChecksum(options);
        const diskFile = await storage.create(request, { ...metafile });

        expect(diskFile).toMatchSnapshot();

        const file = await storage.write({ ...diskFile, body: Readable.from("01234"), start: 0 });

        expect(file).toMatchSnapshot();
    });

    it("diskStorageWithChecksum.delete deletes file and metadata", async () => {
        const storage = new DiskStorageWithChecksum(options);
        const diskFile = await storage.create(request, { ...metafile });

        await storage.write({ ...diskFile, body: Readable.from("01234"), start: 0 });

        const deletedFiles = await storage.delete({ id: diskFile.id });

        expect(deletedFiles).toEqual({
            ...diskFile,
            hash: {
                algorithm: "sha1",
                value: "897988093208097ce65f78fd43e99208926103ea",
            },
            status: "deleted",
        });
        await expect(() => storage.getMeta(diskFile.id)).rejects.toThrow("Not found");
    });
});
