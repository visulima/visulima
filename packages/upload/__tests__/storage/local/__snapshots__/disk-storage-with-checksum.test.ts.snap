// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiskStorageWithChecksum should support checksum resume from fs 1`] = `
File {
  "bytesWritten": 0,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": "2022-02-02T01:00:00.000Z",
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "anonymous/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "created",
}
`;

exports[`DiskStorageWithChecksum should support checksum resume from fs 2`] = `
{
  "bytesWritten": 5,
  "contentType": "video/mp4",
  "createdAt": "2022-02-02T00:00:00.000Z",
  "expiredAt": "2022-02-02T01:00:00.000Z",
  "hash": {
    "algorithm": "sha1",
    "value": "11904a4e8b77f6242e2d288705023adad00a9310",
  },
  "id": "e8fed598250d10ea-7f59007b4b7cf67-120941ca7dc37b78",
  "metadata": {
    "custom": "",
    "lastModified": 1635398061454,
    "mimeType": "video/mp4",
    "name": "testfile.mp4",
    "sha1": "ZAPAntzKARqtb+j3B529GAOf3kI=",
    "size": 64,
  },
  "name": "anonymous/testfile.mp4",
  "originalName": "testfile.mp4",
  "size": 64,
  "status": "part",
}
`;
